### PR TITLE
Generalizes PtySession to work with any Stream type

### PR DESCRIPTION
**Pull request by thomasantony**
I wanted to use `rexpect` with a REPL that is presented over a raw `TcpStream` (and possibly a serial link in the future). The changes I have made creates a generic `StreamSession` struct that can take something that implements `Read` and something that implements `Write` instead of being forced to use a command-line program.

I have also updated `PtySession` to use this. One drawback is that "exp()" no longer gives extra context on error. I am open to suggestions. Thanks!

### CODE REVIEW

**Comment by philippkeller** [comment]
I'm currently sick in bed with stomach pains, so I'll look later into this. I didn't quite understand your use case. I don't understand "repl that is presented over a raw TcpStream". Can you explain this?

**Comment by thomasantony** [comment]
I have been using this to interact with a device that presents a console over a serial port. We send commands and read responses/prompts from the device over this interface. In our particular use case, we have a separate TCP bridge program that passes through data from a TCP port to the serial port and back. We connect to it using a `TCPStream` and I use `StreamSession` to send commands and read responses from the device.

`StreamSession` would also be useful for using `rexpect` with a direct serial port connection. A rust library like https://docs.rs/serialport/3.3.0/serialport also implements the `Read` and `Write` traits and `StreamSession` could work with it. 

Basically, `StreamSession` makes it so you can interact with any object that implements the `Read` and `Write` traits using `rexpect` regardless of the underlying implementation.

**Comment by philippkeller** [comment]
I just checked about the "less context on error", and the loss of information is minimal. When giving a wrong argument to ftp. Before PR:

```
thread 'main' panicked at 'ftp job failed with EOF (End of File): Expected Regex: "Name \(.*\):" but got EOF after reading "ftp: invalid option -- 'l'
Try 'ftp --help' or 'ftp --usage' for more information.
", process terminated with "Exited(Pid(24427), 64)"', examples/ftp.rs:24:33
```

after PR:

```
thread 'main' panicked at 'ftp job failed with EOF (End of File): Expected Regex: "Name \(.*\):" but got EOF after reading "ftp: invalid option -- 'l'
Try 'ftp --help' or 'ftp --usage' for more information.
", process terminated with "unknown"', examples/ftp.rs:24:33
```

=> seems ok to me, as pid and exit code are usually not crucial for debugging

**Comment by philippkeller** [comment]
thanks @thomasantony for this pull PR. Elegantly solved and now allows also sessions over TCP, I'll add an example to master branch (didn't see any way I can add a file in your PR, I only could edit the changes…)
